[git-webkit] Revert's bug filing should be less manual

briannafan opened the pull request:
#### 29338cd2b8a9aa9c7537b3c74944400f9b3ed19f
<pre>
[git-webkit] Revert&apos;s bug filing should be less manual
<a href="https://bugs.webkit.org/show_bug.cgi?id=265593">https://bugs.webkit.org/show_bug.cgi?id=265593</a>
<a href="https://rdar.apple.com/problem/118996008">rdar://problem/118996008</a>

Reviewed by Jonathan Bedard.

Prompts user for a reason for revert and creates revert bug using existing bug details.
If there are multiple commits being reverted and the categories differ, the user is prompted.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.get_commit_info):
(Revert):
(Revert.get_issue_info):
(Revert.create_revert_commit_msg):
(Revert.revert_commit):
(Revert.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
(TestRevert.test_github):
(TestRevert.test_github_two_step):
(TestRevert.test_args):
(test_update):

Canonical link: <a href="https://commits.webkit.org/271939@main">https://commits.webkit.org/271939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1bf90ca10daf1cb1c26afc7952db78e238054d88

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8791 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31440 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32630 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27241 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30801 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10991 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6036 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/30426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/7378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/25695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/6303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/30281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/26788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33967 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/27471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/27221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/6404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/29912 "Failed to checkout and rebase branch from PR 21536") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/8154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7158 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3885 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -73,23 +73,61 @@ def parser(cls, parser, loggers=None):
+                return None, None, None
+            commit_objects.append(commit)
+            commits_to_revert.append(commit.hash)
+            commit_bugs += [Tracker.from_string(i.link) for i in commit.issues if isinstance(Tracker.from_string(i.link).tracker, bugzilla.Tracker)]
```
I would de-duplicate these bugs. It's going to be pretty common to do a multi-revert of a group of commits which reference the same set of bugs.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -73,23 +73,61 @@ def parser(cls, parser, loggers=None):
         # Using --revert-issue to differentiate from PR arg --issue/-i
         if not args.revert_issue:
-            prompt = 'Enter issue URL or title of new issue for the revert: '
+            prompt = 'Enter issue URL or reason for the revert: '
```
This is great! I would also support a `--why` that auto-fills this prompt. That's probably also going to make the code a bit more clear, you're mixing the "issue" object with a "revert reason" currently.

briannafan replied:
I think `--reason` sounds semantically better, but I agree that it's unclear. I also got rid of `--revert-issue` in favour of simply using `--issue`, which I believe now actually works due to previous changes I made.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -73,23 +73,61 @@ def parser(cls, parser, loggers=None):
             issue = Tracker.instance().create(
                 title=args.issue,
-                description=Terminal.input('Issue description: '),
+                description='Revert {} because {}.'.format(string_utils.join(commits), args.issue),
```
See the discussion of "issue" vs "why a few lines up.

briannafan replied:
I think the reason I use `args.issue` after line 98 is to avoid writing code conditioning on whether the terminal input was an issue or a reason. Does that make sense?
Unless, there's a case in which someone enters a bug URL which doesn't exist... Then we would want to re-prompt them for a reason. I can see that being a good justification to separate the two. Let me know if you think this case is realistic.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
```diff
@@ -59,10 +59,10 @@ def parser(cls, parser, loggers=None):
         parser.add_argument(
-            '--revert-issue',
-            dest='revert_issue',
+            '--reason', '-re',
```
Why not just `-r`?

briannafan replied:
`-r` conflicts with the args for issue/radar/bug :(

JonWBedard replied:
Ah! Then we're better off just dropping this. We don't have precedence for a 2 character short argument like this. Another thing you could do (see `git-webkit clone`) is have `reason` be the primary argument, but also respect `--why` and `-w`. That doesn't change your existing code, just the possible arguments a user could provide on the command line.

briannafan committed:
[git-webkit] Revert's bug filing should be less manual
https://bugs.webkit.org/show_bug.cgi?id=265593
rdar://problem/118996008

Reviewed by Jonathan Bedard.

Prompts user for a reason for revert and creates revert bug using existing bug details.
If there are multiple commits being reverted and the categories differ, the user is prompted.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py:
(Revert.get_commit_info):
(Revert):
(Revert.get_issue_info):
(Revert.create_revert_commit_msg):
(Revert.revert_commit):
(Revert.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/revert_unittest.py:
(TestRevert.test_github):
(TestRevert.test_github_two_step):
(TestRevert.test_args):
(test_update):

Canonical link: https://commits.webkit.org/271939@main
